### PR TITLE
Container detail page

### PR DESCRIPTION
**Pull request by ZitaNemeckova**
https://issues.redhat.com/browse/AAH-250

TODO:
- [x] Add tabs
- [x] Detail tab - README editor and display
- [x] Activity tab
- [x] Activity table
- [x] Heat map - followup PR
- [x] Instructions
- [x] Images tab/table
- [x] Empty states
- [x] Images API
- [x] Activity API
- [x] Readme API
- [x] truncate image digest correctly
- [x] Instruction for container registry
- [x] Instructions for image

Screenshot 2021-03-30 at 15 10 52
Screenshot 2021-03-30 at 15 12 16
Screenshot 2021-03-30 at 15 11 04

EMPTY STATES:
Screenshot 2021-03-12 at 14 50 56
Screenshot 2021-03-12 at 14 35 16
Screenshot 2021-03-12 at 14 35 23

### CODE REVIEW

**Comment by sbuenafe-rh** [comment]
@ZitaNemeckova on the header there should be an **Edit** button (to edit the repo). The description should be place in a tooltip (black bkg). 

I'll look into the ReadMe empty state message and button placement. But for the time being, I don't believe you need the "Please add README" text above the Save & Cancel action buttons.

**Comment by ZitaNemeckova** [comment]
@sbuenafe-rh Thanks for review :) I want to add Edit button with a modal form in followup PR to make it easier to review codewise.

**Comment by sbuenafe-rh** [comment]
@ZitaNemeckova tooltip looks good! thanks for the note on the other Edit button with modal. I'll try to draft up an empty state message for the readme.

File: src/components/markdown-editor/markdown-editor.scss [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+    .preview {
+      overflow: scroll;
+      height: 500px;
+      border: 1px solid #ededed;
+      padding: 5px;
+    }
+  }
+
+  .editor {
+    // I don't know why, but just setting width causes this element to get
```
Comment (himdel) [inline src/components/markdown-editor/markdown-editor.scss:25]: would `flex-shrink: 0` help? Sounds like this is caused by flexbox treating width as the optimum, not a constraint.

File: src/components/namespace-form/resources-form.tsx [hunk 2]
```diff
@@ -30,44 +27,22 @@ export class ResourcesForm extends React.Component<IProps, {}> {
...
+          'You can can customize the Resources tab on your profile by entering custom markdown here.'
+        }
+        updateText={value => this.updateResources(value)}
+        editing={true}
+      />
     );
   }
 
   private updateResources(data) {
+    console.log('Data: ' + data);
```
Comment (himdel) [inline src/components/namespace-form/resources-form.tsx:43]: console log (and the comment on line 73?)

File: src/components/markdown-editor/markdown-editror.tsx [hunk 3]
```diff
@@ -0,0 +1,50 @@
+import * as React from 'react';
```
Comment (himdel) [inline src/components/markdown-editor/markdown-editror.tsx:1]: editror -> editor (in the filename)

File: src/components/index.ts [hunk 4]
```diff
@@ -53,3 +53,4 @@ export { EmptyStateFilter } from './empty-state/empty-state-filter';
 export { EmptyStateUnauthorized } from './empty-state/empty-state-unauthorized';
 export { EmptyStateNoData } from './empty-state/empty-state-no-data';
 export { EmptyStateCustom } from './empty-state/empty-state-custom';
+export { MarkdownEditor } from './markdown-editor/markdown-editror';
```
Comment (himdel) [inline src/components/index.ts:56]: editor

File: src/containers/execution-environment-detail/execution_environment_detail.tsx [hunk 5]
```diff
@@ -0,0 +1,580 @@
...
+      },
+      markdownEditing: false,
+    };
+  }
+
+  componentDidMount() {
+    this.queryImages(this.state.container.name);
+    this.queryActivities(this.state.container.name);
+    this.queryReadme(this.state.container.name);
+    this.setState({ loading: false });
```
Comment (himdel) [inline src/containers/execution-environment-detail/execution_environment_detail.tsx:92]: unlikely to be true at this point?

File: src/containers/execution-environment-detail/execution_environment_detail.tsx [hunk 6]
```diff
@@ -0,0 +1,580 @@
...
+          </ClipboardCopy>
+        </td>
+      </tr>
+    );
+  }
+
+  queryReadme(name) {
+    this.setState({ loading: true }, () =>
+      ExecutionEnvironmentAPI.readme(name).then(result => {
+        console.log(result);
```
Comment (himdel) [inline src/containers/execution-environment-detail/execution_environment_detail.tsx:434]: console.log

File: src/containers/execution-environment-detail/execution_environment_detail.tsx (hunk 5 above)
Comment (himdel) [reply src/containers/execution-environment-detail/execution_environment_detail.tsx:92]: and you have an ordering problem here.. even without this line, the *first* query* request to complete will reset loading to false, not the last one ... may need 3 separate loading vars, or one method that deals with all the requests and sets state accordingly

File: src/containers/execution-environment-detail/execution_environment_detail.tsx [hunk 7]
```diff
@@ -0,0 +1,580 @@
...
+import { Paths } from '../../paths';
+import { ImagesAPI, ActivitiesAPI, ExecutionEnvironmentAPI } from '../../api';
+import { pickBy } from 'lodash';
+import * as moment from 'moment';
+import './execution-environment-detail.scss';
+import { TagIcon } from '@patternfly/react-icons';
+
+interface IState {
+  loading: boolean;
+  container: any;
```
Comment (himdel) [inline src/containers/execution-environment-detail/execution_environment_detail.tsx:44]: `container: { name: string; };` ?

File: src/containers/execution-environment-detail/execution_environment_detail.tsx [hunk 8]
```diff
@@ -0,0 +1,580 @@
...
+                  }}
+                >
+                  Edit
+                </Button>
+              )}
+            </Title>
+            {!this.state.markdownEditing && !this.state.readme ? (
+              <EmptyStateNoData
+                title={'No README'}
+                description={'Add README file by using RAW MAkdown editor'}
```
Comment (himdel) [inline src/containers/execution-environment-detail/execution_environment_detail.tsx:173]: MAkdown -> Markdown

File: src/containers/execution-environment-detail/execution_environment_detail.tsx [hunk 9]
```diff
@@ -0,0 +1,580 @@
...
+                  >
+                    Add
+                  </Button>
+                }
+              />
+            ) : (
+              <MarkdownEditor
+                text={this.state.readme}
+                placeholder={
+                  this.state.markdownEditing ? 'Here goes README' : ''
```
Comment (himdel) [inline src/containers/execution-environment-detail/execution_environment_detail.tsx:187]: Here goes Johnny :) .. I guess this one may not be intential? :)

File: src/containers/execution-environment-detail/execution_environment_detail.tsx [hunk 10]
```diff
@@ -0,0 +1,580 @@
...
+              <table aria-label='Activities' className='pf-c-table'>
+                <SortTable
+                  options={{
+                    headers: [
+                      { title: 'Change', type: 'none', id: 'change' },
+                      { title: 'Date', type: 'none', id: 'date' },
+                    ],
+                  }}
+                  params={params}
+                  updateParams={() => {}}
```
Comment (himdel) [inline src/containers/execution-environment-detail/execution_environment_detail.tsx:255]: intentionally not resortable?

File: src/containers/execution-environment-detail/execution_environment_detail.tsx [hunk 11]
```diff
@@ -0,0 +1,580 @@
...
+                this.queryImages(this.state.container.name),
+              )
+            }
+            count={this.state.images.length}
+            isTop
+          />
+        </div>
+        <div>
+          <AppliedFilters
+            updateParams={() => console.log('update params and queryImages')}
```
Comment (himdel) [inline src/containers/execution-environment-detail/execution_environment_detail.tsx:364]: console log / TODO?

File: src/containers/execution-environment-detail/execution_environment_detail.tsx [hunk 12]
```diff
@@ -0,0 +1,580 @@
...
+  private renderTableRow(image: any, index: number) {
+    let instruction =
+      image.tags.length === 0
+        ? image.digest
+        : this.state.container.name + ':' + image.tags[0];
+    return (
+      <tr key={index}>
+        <td>
+          {image.tags.map(tag => (
+            <Label variant='outline' icon={<TagIcon />}>
```
Comment (himdel) [inline src/containers/execution-environment-detail/execution_environment_detail.tsx:407]: missing key?

File: src/containers/execution-environment-detail/execution_environment_detail.tsx [hunk 13]
```diff
@@ -0,0 +1,580 @@
...
+      }),
+    );
+  }
+
+  queryImages(name) {
+    this.setState({ loading: true }, () =>
+      ImagesAPI.list(name).then(result => {
+        let images = [];
+        result.data.data.forEach(object => {
+          let image = pickBy(object, function(value, key) {
```
Comment (himdel) [inline src/containers/execution-environment-detail/execution_environment_detail.tsx:459]: is this `let image = pick(object, ['digest', 'tags', 'pulp_created']);`?

File: src/containers/execution-environment-detail/execution_environment_detail.tsx [hunk 14]
```diff
@@ -0,0 +1,580 @@
...
+            </ToolbarContent>
+          </Toolbar>
+          <Pagination
+            params={params}
+            updateParams={p =>
+              this.updateParams(p, () =>
+                this.queryImages(this.state.container.name),
+              )
+            }
+            count={this.state.images.length}
```
Comment (himdel) [inline src/containers/execution-environment-detail/execution_environment_detail.tsx:358]: does the API return `meta.count` for images? Giving the same count as the number of loaded images means no pagination ever, if the number of loaded images depends on items per page.

File: src/containers/execution-environment-detail/execution_environment_detail.tsx [hunk 15]
```diff
@@ -0,0 +1,580 @@
...
+        </table>
+        <div style={{ paddingTop: '24px', paddingBottom: '8px' }}>
+          <Pagination
+            params={params}
+            updateParams={p =>
+              this.updateParams(p, () =>
+                this.queryImages(this.state.container.name),
+              )
+            }
+            count={this.state.images.length}
```
Comment (himdel) [inline src/containers/execution-environment-detail/execution_environment_detail.tsx:391]: (and here, `meta.count`)

File: src/containers/execution-environment-detail/execution_environment_detail.tsx (hunk 10 above)
Comment (ZitaNemeckova) [reply src/containers/execution-environment-detail/execution_environment_detail.tsx:255]: Yes. API is not able to sort or filter activities right now.

File: src/containers/execution-environment-detail/execution_environment_detail.tsx (hunk 11 above)
Comment (ZitaNemeckova) [reply src/containers/execution-environment-detail/execution_environment_detail.tsx]: yep, fixed :)

File: src/containers/execution-environment-detail/execution_environment_detail.tsx (hunk 5 above)
Comment (ZitaNemeckova) [reply src/containers/execution-environment-detail/execution_environment_detail.tsx]: Fixed :)

Comment (himdel) [reply src/containers/execution-environment-detail/execution_environment_detail.tsx]: Nice :) except now the query* calls from buttons won't unset loading..

File: src/containers/execution-environment-detail/execution_environment_detail.tsx [hunk 16]
```diff
@@ -0,0 +1,591 @@
...
+                  }}
+                >
+                  Edit
+                </Button>
+              )}
+            </Title>
+            {!this.state.markdownEditing && !this.state.readme ? (
+              <EmptyStateNoData
+                title={'No README'}
+                description={'Add README file by using RAW Makdown editor'}
```
Comment (newswangerd) [inline src/containers/execution-environment-detail/execution_environment_detail.tsx:180]: Lets rephrase this to "Add a README with instructions for using this container."

File: src/containers/execution-environment-detail/execution_environment_detail.tsx [hunk 17]
```diff
@@ -0,0 +1,591 @@
...
+
+    return (
+      <Flex direction={{ default: 'column' }}>
+        <FlexItem>
+          <Section className='body card-area'>
+            {' '}
+            <Title headingLevel='h2' size='lg'>
+              Instructions
+            </Title>
+            <ClipboardCopy isCode isReadOnly variant={'expansion'}>
```
Comment (newswangerd) [inline src/containers/execution-environment-detail/execution_environment_detail.tsx:156]: Don't use `isCode` here.

File: src/containers/execution-environment-detail/execution_environment_detail.tsx [hunk 18]
```diff
@@ -0,0 +1,591 @@
...
+  renderDetail() {
+    const instructions = 'podman pull ' + this.state.container.name + ':latest';
+
+    return (
+      <Flex direction={{ default: 'column' }}>
+        <FlexItem>
+          <Section className='body card-area'>
+            {' '}
+            <Title headingLevel='h2' size='lg'>
+              Instructions
```
Comment (newswangerd) [inline src/containers/execution-environment-detail/execution_environment_detail.tsx:154]: Maybe change this to "Pull this image"?

"Instructions" is so vague.

File: src/containers/execution-environment-detail/execution_environment_detail.tsx [hunk 19]
```diff
@@ -0,0 +1,591 @@
...
+            </Title>
+            <ClipboardCopy isCode isReadOnly variant={'expansion'}>
+              {instructions}
```
Comment (newswangerd) [inline src/containers/execution-environment-detail/execution_environment_detail.tsx:157]: Same as with the images page: the command has to have the registry host in it

File: src/containers/execution-environment-detail/execution_environment_detail.tsx [hunk 20]
```diff
@@ -0,0 +1,591 @@
...
+              <Tabs
+                tabs={tabs}
+                params={params}
+                updateParams={p => this.updateParams(p)}
+              />
+            </div>
+          </div>
+        </BaseHeader>
+        <Main>
+          {params['tab'] === 'detail' && this.renderDetail()}
```
Comment (newswangerd) [inline src/containers/execution-environment-detail/execution_environment_detail.tsx:137]: This component needs to be split up into 3 components, one for each tab. Putting all the logic into one component makes it very difficult to find things and figure out what's happening.

Additionally, each tab should have it's own URL route, just like we do for collection details.

I would recommend the following routing
- `containers/<name>/` -> Detail
- `containers/<name>/_content/activity/` -> Activity
- `containers/<name>/_content/images/` -> Images

The reason for  `_content` is because we have to route urls for `containers/<name>/` and `containers/<namespace>/<name>/`, but `_content` isn't a valid docker image name, so that avoids any potential routing conflicts between our routes and the container names.

File: src/containers/execution-environment-detail/execution_environment_detail.tsx [hunk 21]
```diff
@@ -0,0 +1,591 @@
...
+                </Button>
+              </React.Fragment>
+            )}
+          </Section>
+        </FlexItem>
+      </Flex>
+    );
+  }
+
+  renderActivity() {
```
Comment (newswangerd) [inline src/containers/execution-environment-detail/execution_environment_detail.tsx:236]: Lets add a "Load more" link at the bottom of the list of activities that loads the next "page". As you've pointed out, using pagination here doesn't really work because there isn't a 1 to 1 relationship between rows here and API responses. Load more would solve this by just requesting the next page of activity and appending it to the bottom of the table.

File: src/containers/execution-environment-detail/execution_environment_detail.tsx [hunk 22]
```diff
@@ -0,0 +1,591 @@
...
+import { pickBy } from 'lodash';
+import * as moment from 'moment';
+import './execution-environment-detail.scss';
+import { TagIcon } from '@patternfly/react-icons';
+
+interface IState {
+  loading: boolean;
+  container: { name: string };
+  readme: string;
+  images: any[];
```
Comment (newswangerd) [inline src/containers/execution-environment-detail/execution_environment_detail.tsx:45]: Don't use `any`. Need to add types for the api responses for images, container, and activities.

File: src/containers/execution-environment-detail/execution_environment_detail.tsx [hunk 23]
```diff
@@ -0,0 +1,591 @@
...
+          type: 'none',
+          id: 'layers',
+        },
+        {
+          title: 'Size',
+          type: 'none',
+          id: 'size',
+        },
+        {
+          title: 'Image',
```
Comment (newswangerd) [inline src/containers/execution-environment-detail/execution_environment_detail.tsx:319]: Rename this to "Digest"

File: src/containers/execution-environment-detail/execution_environment_detail.tsx [hunk 24]
```diff
@@ -0,0 +1,591 @@
...
+          ))}
+        </td>
+        <Tooltip content={moment(image.pulp_created).format('MMMM Do YYYY')}>
+          <td>{moment(image.pulp_created).fromNow()}</td>
+        </Tooltip>
+        <td>{image.layers}</td>
+        <td>{image.size}</td>
+        <td>
+          <Popover position={PopoverPosition.top} bodyContent={image.digest}>
+            <Label color='blue'>{image.digest.slice(0, 8)}</Label>
```
Comment (newswangerd) [inline src/containers/execution-environment-detail/execution_environment_detail.tsx:432]: I'd recommend `{image.digest.split(':')[0]:image.digest.split(':')[1].slice(0, 8)}`.

That way you always get the digest type and at least 8 digits of the actual digest.

**Review by newswangerd — CHANGES_REQUESTED** [review]

File: src/containers/execution-environment-detail/execution_environment_detail.tsx (hunk 21 above)
Comment (newswangerd) [reply src/containers/execution-environment-detail/execution_environment_detail.tsx:236]: Actually, can you create an issue for this instead? Lets tackle this in a different PR

Comment (ZitaNemeckova) [reply src/containers/execution-environment-detail/execution_environment_detail.tsx:236]: https://issues.redhat.com/browse/AAH-461

File: src/containers/execution-environment-detail/execution_environment_detail.tsx [hunk 25]
```diff
@@ -0,0 +1,591 @@
...
+              README
+              {!this.state.markdownEditing && this.state.readme && (
+                <Button
+                  className={'edit-button'}
+                  variant={'primary'}
+                  onClick={() => {
+                    this.setState({ markdownEditing: true });
+                  }}
+                >
+                  Edit
```
[the remark this answers is not in the document]
Comment (ZitaNemeckova) [reply src/containers/execution-environment-detail/execution_environment_detail.tsx:135]: Tracking via https://issues.redhat.com/browse/AAH-467

File: src/containers/execution-environment-detail/execution_environment_detail.tsx [hunk 26]
```diff
@@ -0,0 +1,591 @@
...
+    }
+
+    this.state = {
+      loading: true,
+      images: [],
+      numberOfImages: 0,
+      activities: [],
+      container: { name: this.props.match.params['container'] },
+      readme: '',
+      params: {
```
[the remark this answers is not in the document]
Comment (newswangerd) [reply src/containers/execution-environment-detail/execution_environment_detail.tsx]: This still isn't working. The parameters are in the URL, but they don't get added to the filter when you load the page.

File: src/containers/execution-environment-detail/execution_environment_detail.tsx [hunk 27]
```diff
@@ -0,0 +1,223 @@
...
+          </div>
+        </BaseHeader>
+        <Main>{this.renderDetail()}</Main>
+      </React.Fragment>
+    );
+  }
+
+  renderDetail() {
+    const instructions =
+      'podman pull localhost:5001/' + this.state.container.name + ':latest';
```
Comment (newswangerd) [inline src/containers/execution-environment-detail/execution_environment_detail.tsx:111]: Don't hardcode the URL into the podman pull instructions. Use the URL from the browser.

File: src/containers/execution-environment-detail/execution_environment_detail.tsx [hunk 28]
```diff
@@ -0,0 +1,223 @@
...
+        <Redirect
+          to={formatPath(Paths.executionEnvironmentDetailImages, {
+            container: this.state.container.name,
+          })}
+        />
+      );
+    }
+    return (
+      <React.Fragment>
+        <BaseHeader
```
Comment (newswangerd) [inline src/containers/execution-environment-detail/execution_environment_detail.tsx:76]: Since the header is the same across all the EE pages, it would be wise to create one component with the shared information that can be used across all the pages.

File: src/containers/execution-environment-detail/execution_environment_detail.tsx [hunk 29]
```diff
@@ -0,0 +1,223 @@
...
+  }
+
+  componentDidMount() {
+    const { container } = this.state;
+    this.queryReadme(container.name);
+  }
+
+  render() {
+    const tabs = ['Detail', 'Activity', 'Images'];
+    const description = '';
```
Comment (newswangerd) [inline src/containers/execution-environment-detail/execution_environment_detail.tsx:55]: Description should be available from the API.

File: src/paths.ts [hunk 30]
```diff
@@ -15,6 +15,9 @@ export function formatPath(path: Paths, data: any, params?: object) {
 }
 
 export enum Paths {
+  executionEnvironmentDetailActivities = '/containers/:container/_content/activity',
```
Comment (newswangerd) [inline src/paths.ts:18]: All these routes need to match `/containers/:container/` and `/container/:container/:namespace/`

File: src/containers/execution-environment-detail/execution_environment_detail.tsx (hunk 22 above)
Comment (newswangerd) [reply src/containers/execution-environment-detail/execution_environment_detail.tsx]: This still needs to be fixed.

File: src/containers/execution-environment-detail/execution_environment_detail.tsx [hunk 31]
```diff
@@ -0,0 +1,591 @@
...
+            <Label variant='outline' key={tag} icon={<TagIcon />}>
+              {tag}
+            </Label>
+          ))}
+        </td>
+        <Tooltip content={moment(image.pulp_created).format('MMMM Do YYYY')}>
+          <td>{moment(image.pulp_created).fromNow()}</td>
+        </Tooltip>
+        <td>{image.layers}</td>
+        <td>{image.size}</td>
```
[the remark this answers is not in the document]
Comment (newswangerd) [reply src/containers/execution-environment-detail/execution_environment_detail.tsx]: This still needs to be fixed

**Review by newswangerd — CHANGES_REQUESTED** [review]

File: src/containers/execution-environment-detail/execution_environment_detail.tsx (hunk 27 above)
Comment (ZitaNemeckova) [reply src/containers/execution-environment-detail/execution_environment_detail.tsx:111]: Not sure if it's correct given I hardcode change of ports.

File: src/api/activities.ts [hunk 32]
```diff
@@ -0,0 +1,15 @@
...
+
+class API extends HubAPI {
+  apiPath = this.getUIPath('execution-environments/repositories/');
+
+  constructor() {
+    super();
+  }
+
+  list(id) {
+    return this.http.get(this.apiPath + id + '/_content/history/');
```
[the remark this answers is not in the document]
Comment (ZitaNemeckova) [reply src/api/activities.ts]: Ok, but there's no pagination for now as UI is parsing the data and it's not one to one.

File: src/containers/execution-environment-detail/execution_environment_detail_activities.tsx [hunk 33]
```diff
@@ -0,0 +1,265 @@
...
+          <Tooltip content={description}>
+            <p className={'truncated'}>{description}</p>
+          </Tooltip>
+          <span />
+          <div className='tab-link-container'>
+            <div className='tabs'>
+              <Tabs
+                tabs={tabs}
+                params={{ tab: 'activity' }}
+                updateParams={p => this.setState({ redirect: p.tab })}
```
[the remark this answers is not in the document]
Comment (ZitaNemeckova) [reply src/containers/execution-environment-detail/execution_environment_detail_activities.tsx:92]: To be fixed later. Tracker https://issues.redhat.com/browse/AAH-477

File: src/containers/execution-environment-detail/execution_environment_detail.tsx [hunk 34]
```diff
@@ -0,0 +1,208 @@
...
+      redirect: null,
+    };
+  }
+
+  componentDidMount() {
+    this.queryReadme(this.props.match.params['container']);
+  }
+
+  render() {
+    if (this.state.redirect === 'activity') {
```
Comment (newswangerd) [inline src/containers/execution-environment-detail/execution_environment_detail.tsx:52]: This should be moved into the header.

File: src/components/execution-environment-header/execution-environment-header.tsx [hunk 35]
```diff
@@ -0,0 +1,75 @@
...
+interface IState {
+  container: { name: string; description: string; namespace: string };
+}
+
+export class ExecutionEnvironmentHeader extends React.Component<
+  IProps,
+  IState
+> {
+  componentDidMount() {
+    ExecutionEnvironmentAPI.get(this.props.id)
```
Comment (newswangerd) [inline src/components/execution-environment-header/execution-environment-header.tsx:22]: Components should never make API calls unless it's absolutely necessary to do so.

File: src/containers/execution-environment-detail/execution_environment_detail.tsx [hunk 36]
```diff
@@ -0,0 +1,223 @@
...
+  Button,
+  Tooltip,
+} from '@patternfly/react-core';
+import { formatPath, Paths } from '../../paths';
+import { ExecutionEnvironmentAPI } from '../../api';
+import './execution-environment-detail.scss';
+
+interface IState {
+  loading: boolean;
+  container: { name: string };
```
Comment (newswangerd) [inline src/containers/execution-environment-detail/execution_environment_detail.tsx]: This should use a container type that gets imported from `api/response-types`, especially since this type is used on all the container detail views as well as the container list view.

**Review by newswangerd — APPROVED** [review]
There are still a few things here that violate best practices, which will make updating and debugging this page later more challenging. However, in the interest of finishing this feature, I'm going to mark this approved and take it on as technical debt for later.
